crates/ruff_python_formatter/src/verbatim.rs

```diff
@@ -573,33 +596,275 @@ impl Format<PyFormatContext<'_>> for TrailingFormatOffComment<'_> {
+                    // It, therefore, feels wrong to add additional complexity to the very hot `Printer::print_char` function,
+                    // to work around this special case. Therefore, work around the Printer behavior here, in the cold verbatim-formatting.
+                    if f.context().source()[trimmed_line_range].width() == 0 {
+                        empty_line().fmt(f)?;
+                    } else {
+                        hard_line_break().fmt(f)?;
+                    }
+                }
+            }
+
+            first = false;
+        }
+
+        Ok(())
+    }
+}
+
+struct LogicalLinesIter<I> {
+    lexer: I,
+    // The end of the last logical line
+    last_line_end: TextSize,
+    // The position where the content to lex ends.
+    content_end: TextSize,
+}
+
+impl<I> LogicalLinesIter<I> {
+    fn new(lexer: I, verbatim_range: TextRange) -> Self {
+        Self {
+            lexer,
+            last_line_end: verbatim_range.start(),
+            content_end: verbatim_range.end(),
+        }
+    }
+}
+
+impl<I> Iterator for LogicalLinesIter<I>
+where
+    I: Iterator<Item = LexResult>,
+{
+    type Item = FormatResult<LogicalLine>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let mut parens = 0u32;
+        let mut contains_newlines = ContainsNewlines::No;
+
+        let (content_end, full_end) = loop {
+            match self.lexer.next() {
+                Some(Ok((token, range))) => match token {
+                    Tok::Newline => break (range.start(), range.end()),
+                    // Ignore if inside an expression
+                    Tok::NonLogicalNewline if parens == 0 => break (range.start(), range.end()),
+                    Tok::NonLogicalNewline => {
+                        contains_newlines = ContainsNewlines::Yes;
+                    }
+                    Tok::Lbrace | Tok::Lpar | Tok::Lsqb => {
+                        parens = parens.saturating_add(1);
+                    }
+                    Tok::Rbrace | Tok::Rpar | Tok::Rsqb => {
+                        parens = parens.saturating_sub(1);
+                    }
+                    Tok::String { value, .. } if value.contains(['\n', '\r']) => {
+                        contains_newlines = ContainsNewlines::Yes;
+                    }
+                    _ => {}
+                },
+                None => {
+                    // Is there any remaining content after the last token? If so, return it as the last
+                    // logical line.
+                    return if self.last_line_end < self.content_end {
+                        let content_start = self.last_line_end;
+                        self.last_line_end = self.content_end;
+                        Some(Ok(LogicalLine {
+                            content_range: TextRange::new(content_start, self.content_end),
+                            contains_newlines,
+                            has_trailing_newline: false,
```

Comment:
what would that content be and how do we know whether it contains newlines?

Reply:
I extended the comment to mention what content it is and changed `contains_newlines` to always be `No` because the lexer would otherwise return a `Newline` token.